linera-core/src/worker.rs

```diff
@@ -1061,10 +1077,15 @@ where
                 value_hash: blobs[0].hash(),
             }
         );
+
+        let certificate_for_logging = certificate.clone();
+        let mut duplicated = false;
         let (info, actions) = match certificate.value() {
             CertificateValue::ValidatedBlock { .. } => {
                 // Confirm the validated block.
-                self.process_validated_block(certificate).await?
+                let result = self.process_validated_block(certificate).await?;
+                duplicated = result.2;
+                (result.0, result.1)
```

Comment:
Also not pretty, but I'd still prefer this:

```suggestion
                let (info, actions, d) = self.process_validated_block(certificate).await?;
                duplicated = d;
                (info, actions)
```